packages/kit/types/page.d.ts

```diff
@@ -1,27 +1,42 @@
-export type LoadInput = {
+import { MaybePromise } from './helper';
+
+export type LoadInput<Context extends Record<string, any> = Record<string, any>> = {
 	page: Page;
 	fetch: (info: RequestInfo, init?: RequestInit) => Promise<Response>;
 	session: any;
-	context: Record<string, any>;
+	context: Context;
 };
 
-export type ErrorLoadInput = LoadInput & {
+export type ErrorLoadInput<
+	Context extends Record<string, any> = Record<string, any>
+> = LoadInput<Context> & {
 	status: number;
 	error: Error;
 };
 
-export type LoadOutput = {
+export type LoadOutput<
+	Props extends MaybePromise<Record<string, any>> = MaybePromise<Record<string, any>>,
+	Context extends Record<string, any> = Record<string, any>
+> = {
 	status?: number;
 	error?: string | Error;
 	redirect?: string;
-	props?: Record<string, any> | Promise<Record<string, any>>;
-	context?: Record<string, any>;
+	props?: Props;
+	context?: Context;
 	maxage?: number;
 };
 
-/* Publicized Types */
-export type Load = (input: LoadInput) => LoadOutput | Promise<LoadOutput>;
-export type ErrorLoad = (input: ErrorLoadInput) => LoadOutput | Promise<LoadOutput>;
+// Publicized Types
+export type Load<
+	InputContext extends Record<string, any> = Record<string, any>,
+	OutputContext extends Record<string, any> = Record<string, any>,
+	Props extends MaybePromise<Record<string, any>> = MaybePromise<Record<string, any>>
+> = (input: LoadInput<InputContext>) => MaybePromise<LoadOutput<Props, OutputContext>>;
+export type ErrorLoad<
+	InputContext extends Record<string, any> = Record<string, any>,
+	OutputContext extends Record<string, any> = Record<string, any>,
+	Props extends MaybePromise<Record<string, any>> = MaybePromise<Record<string, any>>
+> = (input: ErrorLoadInput<InputContext>) => MaybePromise<LoadOutput<Props, OutputContext>>;
```

Comment:
I'm not sure about putting `Props` as the last one, it's arguably the most frequently used one, but it might not even matter too since we usually type them in `<script>` tag itself. Still, this might be useful when (and if possible) the exported props from `<script>` could read and check the types from return object of `load` and vice-versa.

3 positional type arguments might also be a handful to type every time, especially when they have no connections to each other. We could also just replace it altogether for named type arguments. That way, any of the types can be passed and the others can be skipped altogether.

Reply:
To guard against breaking changes it's probably the easiest to only have one argument. I think it was more in the past but was changed to one. I'm also split on what the most frequent used here is. If people don't want to type it all, they can always create custom types like `type MyLoad<Props> = Load<Record<string, any>, Record<string, any>, Props>` .

Reply:
If what we want is to only have one argument, then named type argument might be the choice for it. I've tried around with this and works just the same, but this is arguably unconventional in some cases.

```ts
export type Load<
  T extends { input?: object; output?: object; props?: MaybePromise<object> } = any
> = (input: LoadInput<T['input']>) => MaybePromise<LoadOutput<T['props'], T['output']>>;

/* all of this works */
export const load: Load<{
  input: { entry: string };
  output: { hello: string };
  props: { data: { name: string } };
}> = async ({ fetch, context }) => {
  // ...
  return {
    props: {
      data: { name: '' }
    },
    context: { hello: 'world' },
  };
};
// only input
export const load: Load<{ input: { entry: string } }> = async ({ fetch, context }) => {
  // ...
};
// only output
export const load: Load<{ output: { hello: string } }> = async ({ fetch, context }) => {
  // ...
};
// only props
export const load: Load<{ props: { data: { name: string } } }> = async ({ fetch, context }) => {
  // ...
};
```